pyclaragenomics/claragenomics/io/pafio.py

```diff
@@ -1,117 +1,169 @@
-    8	int	Target start on original strand (0-based)
-    9	int	Target end on original strand (0-based)
-    10	int	Number of residue matches
-    11	int	Alignment block length
-    12	int	Mapping quality (0-255; 255 for missing)
+        tags (dict): Dictionary of SAM style tags, keys should match those given
+            in minimap2 docs: https://lh3.github.io/minimap2/minimap2.html
+
+    Returns:
+    str:
+        Tab separated string of tags in format 'tag:type:value'
+        
+    """
+    return "\t".join("{}:{}:{}".format(k, REV_TYPES.get(k), v) for k, v in tags.items())
+
+
+def _record_to_str(rec):
+    """Convert PAF namedtuple to sting for output
+
+    Args:
+        rec (namedtuple): PAF entry namedtuple, generated by read_paf.
+
+    Returns:
+        str: Tab separated string of PAF fields
+
+    """
+    return "\t".join([str(x) for x in rec[:-1]]) + "\t" + _tags_to_str(rec[-1]) + "\n"
+
+
+def _paf_generator(file_like, fields=None):
+    """Generator that yields namedtuples from a PAF file
+
+    Args:
+        file_like (file-like object): Generally an object with a read() method
+            such as sys.stdin, a file handler from open() or io.StringIO.
+        fields (list): List of field names to use for records, must have 13
+            entries.
+
+    Yields:
+        namedtuple: Correctly formatted PAF record and a dict of extra tags.
+
+    Raises:
+        ValueError: If param `fields` does not have 13 entries.
+
+    """
+    if len(fields) != 13:
+        raise ValueError("{} fields provided, expected 13".format(len(fields)))
+    PAF = namedtuple("PAF", fields)
+    for rec in file_like:
+        rec = rec.strip()
+        if not rec:
+            continue
+        rec = rec.split("\t")
+        yield PAF(*(int(x) if x.isdigit() else x for x in rec[:12]), _parse_tags(rec[12:]))
+
+
+def read_paf(filepath, fields=None):
+    """Read a minimap2 PAF file into a list
+
+    Args:
+        filepath (str): Path to read PAF file from
+        fields (list): List of fields to use for records, must have 13 entries.
+            Default: ["query_name", "query_length", "query_start", "query_end",
+                      "strand", "target_name", "target_length", "target_start",
+                      "target_end", "residue_matches", "alignment_block_length",
+                      "mapping_quality", "tags"]
+
+    Returns:
+        list: List of namedtuples
+
+    """
+    if fields is None:
+        fields = [
+            "query_name",
+            "query_length",
+            "query_start",
+            "query_end",
+            "strand",
+            "target_name",
+            "target_length",
+            "target_start",
+            "target_end",
+            "residue_matches",
+            "alignment_block_length",
+            "mapping_quality",
+            "tags",
+        ]
+
+    with open(filepath, "r") as fh:
+        return list(_paf_generator(fh, fields=fields))
+
+
+def write_paf(records, filepath):
```

Comment:
@alexomics one of the use cases is to pass in a list of custom overlaps that can be written out to a paf file. In this API, it looks like we need the PAF namedtuple that is only returned by the `read_paf` function. Could we add the capability to instantiate a PAF namedtuple outside of the read_paf function?

e.g. https://github.com/clara-genomics/ClaraGenomicsAnalysis/blob/dev-v0.5.0/pyclaragenomics/claragenomics/simulators/readsim.py#L81

Reply:
Yes!

Reply:
My latest commits (9a7a4ae, fbbc131) should clear this up.

Reply:
I have rolled back to using the same (hard coded) fields from before my pull req. I have only added a `tags` field at the end. The namedtuple has changed back to `Overlap` and is importable again.